Feature: add Surface_mesh_approximation package.

aboudev opened the pull request:
## Summary of Changes

Adding a new package named Surface_mesh_approximation.

## Release Management

* Affected package(s): add plugin for polyhedron demo
* Issue(s) solved (if any): no
* Feature/Small Feature (if any): [Feature on wiki page](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/VSA)
* Link to compiled documentation (obligatory for small feature) [doc](https://cgal.geometryfactory.com/~sloriot/doc/VSA_3/Surface_mesh_approximation/group__PkgTSMA.html)
* License and copyright ownership: GPL and Inria

sloriot commented:
@aboudev, please review 22453a0 to make sure I did not do something wrong.

aboudev commented:
> @aboudev, please review [22453a0](https://github.com/CGAL/cgal/commit/22453a069d6bf2f3f3077de5e51a64476642eaf3) to make sure I did not do something wrong.

Checked, the parameter `m` in `on_comboMetric_currentIndexChanged(const int m)` is not used and removed.

sloriot commented:
@aboudev you can look at the first test results of your package on [that page](https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-37.shtml).
Tests look good for a first try, only a few errors on some platforms must be fixed.

sloriot commented:
I think I have fixed the errors and warnings.

sloriot commented on Surface_mesh_approximation/doc/Surface_mesh_approximation/NamedParameters.txt:
```diff
@@ -0,0 +1,173 @@
+
+\cgalNPBegin{min_error_drop} \anchor VSA_min_error_drop
+the minimum total error drop to approximate the input mesh.\n
+\b Type : `boost::optional<GeomTraits::FT>` \n
```
if there is a default value, it makes no sense to have an optional, it should be FT.

aboudev replied:
fixed

sloriot commented on Surface_mesh_approximation/doc/Surface_mesh_approximation/NamedParameters.txt:
```diff
@@ -0,0 +1,173 @@
+
+\cgalNPBegin{max_number_of_proxies} \anchor VSA_max_number_of_proxies
+the maximum number of proxies used to approximate the input mesh.\n
+\b Type : `boost::optional<std::size_t>` \n
```
Can't we get rid of this optional too?

aboudev replied:
Fixed.

sloriot replied:
Well it's not only a doc pb, the code also must be updated.

sloriot replied:
I mean everywhere `get_param(internal_np::max_number_of_proxies)` is used (only one place IFAIR).

aboudev replied:
Currently the defaults of `max_number_of_proxies` and `min_error_drop` are used only when neither is given. A parameter is ignored if the other one is provided. We introduce `boost::optional` to describe this kind of non-exist-default situation.

aboudev commented:
> I think I have fixed the errors and warnings.

Thanks!
There seem to be a package dependencies error given by travis every time. I tried but could not fix that.

sloriot commented:
About the dependency error, I pushed something also, I did not see the travis output after so I don't know if it is fixed or not.

lrineau commented:
[**@sloriot**](https://github.com/sloriot) commented on [Nov 7, 2018, 5:13 PM GMT+1](https://github.com/CGAL/cgal/pull/3439#issuecomment-436679466 "2018-11-07T16:13:38Z - Replied by Github Reply Comments"):
> About the dependency error, I pushed something also, I did not see the travis output after so I don't know if it is fixed or not.

That seems fixed, now.

sloriot commented:
@maxGimeno Could you please have a look at the plugin? There is probably an error in the cmake provided the error on windows [here](https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.14-Ic-45/Polyhedron_Demo/TestReport_afabri_x64_Cygwin-Windows10_MSVC2013-Release-64bits.gz) and [there](
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-4.14-Ic-45/Polyhedron_Demo/TestReport_afabri_x64_Cygwin-Windows10_MSVC2017-Release-64bits.gz).

The tests and examples are good.
@palliez Could you make a review of the demo plugin before the branch is merged into master?

maxGimeno commented:
From what I see, exports are missing for Color_cheat_sheet and VSA_wrapper.
header files should have something like:

`#ifdef color_cheat_sheet_EXPORTS`
`#  define COLOR_CHEAT_SHEET_EXPORT Q_DECL_EXPORT`
`#else`
`#  define COLOR_CHEAT_SHEET_EXPORT Q_DECL_IMPORT`
`#endif`
and class declared like this :
`class COLOR_CHEAT_SHEET_EXPORT Color_cheat_sheet`

It is specific to windows, needed by the dll system to know what should be imported or exported from/to a dll.

Appart from that, everything seems fine.

afabri commented:
Hello @aboudev, 
I have several remarks concerning the demo
-  The widget should have stretchers at the bottom, so that its content becomes compact.
-  Are values for error drop, chord error input dependent?
-  I guess you make reasonable choices for metric etc.
-  Do not visualize anchor and patch planes, that is produce the objects but don't make them visible.
-  Describe the workflow (maybe a short section in the User Manual). I guess I have to initialize seeds and can then directly extract the mesh. When I add or teleport a proxy shouldn't it update the extracted mesh? 
- Put the soup-patch objects in an  object group.

afabri commented:
![capture](https://user-images.githubusercontent.com/3263539/48855715-ed8ae880-edb4-11e8-8ed1-b29540d078a6.PNG)

lrineau commented:
@sloriot This PR weights 3.6MB.

afabri commented:
Hello @aboudev, what I still don't understand:  When I add or teleport a proxy  I would expect that something visual changes.

aboudev commented:
Hi @afabri , I added seed face unit normal visualization. So when we add or teleport, the seed placement change can be visualized through their normals. I also added an automatic mesh extraction check box, which allows us to extract mesh at every fit/add/teleport/split step.

aboudev commented:
Hi @maxGimeno , the face color can be set by the `Scene_surface_mesh_item` and its `face_patch_id`. But the color map (id -> color) does not seem to be customizable, because `sm_item->invalidateOpenGLBuffers()` will rewrite the color map every time.

maxGimeno commented:
If you rebase on master you will get access to a very recent feature of the surface_mesh_item: setComputeColorAutomatically() or something like that. If you set that to false it won't overwrite the color map.

maxGimeno commented:
warnings: https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-52.shtml#Surface_mesh_approximation

maxGimeno commented:
There are errors in the exports declarations : the name of the export is composed with the name of the cmake target and the suffix _EXPORTS, and for VSA_wrapper and Color_cheat_sheet, you used respectively vsa_wrapper_EXPORTS and color_cheat_sheet_EXPORTS, when the actual target is called surface_mesh_approximation_plugin.
The fix for windows is to replace those two by surface_mesh_approximation_plugin_EXPORTS in the #ifdef.

aboudev commented:
> If you rebase on master you will get access to a very recent feature of the surface_mesh_item: setComputeColorAutomatically() or something like that. If you set that to false it won't overwrite the color map.

That's really nice. I'm not sure if we are allowed to rebase in a pull request. @sloriot

aboudev commented:
> warnings: https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-52.shtml#Surface_mesh_approximation

This warning C4996 is everywhere and I have no idea how to fix that. I found a [related issue](https://github.com/boostorg/beast/issues/1272), it's probably caused by the `/std:c++latest` flag and `boost::optional`.

maxGimeno commented:
> > If you rebase on master you will get access to a very recent feature of the surface_mesh_item: setComputeColorAutomatically() or something like that. If you set that to false it won't overwrite the color map.
> 
> That's really nice. I'm not sure if we are allowed to rebase in a pull request. @sloriot

You can safely merge master in your branch, that's how we usually fix conflicts.

sloriot commented:
If I convert png to jpg the size goes down to 2.4MB. (FYI, completely removinhg `fig`, brings the size down to 845KB.)
@aboudev @lrineau shall I push that version?

aboudev commented:
> If I convert png to jpg the size goes down to 2.4MB. (FYI, completely removinhg `fig`, brings the size down to 845KB.)
> @aboudev @lrineau shall I push that version?

Sounds good, I'm okay with that.

aboudev commented:
After rebasing to the master, the face color worked as expected.

maxGimeno commented:
Warnings : https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-61.shtml#Polyhedron_Demo
uninitianilzed variables.

sloriot commented:
Those `Wmaybe-unitialized`  warnings may be tedious to fix. I already started a new branch to fix a bunch of them but it is taking a LOT of time to spot each issue. I would suggest not to block the integration if those are the only warnings remaining because as I already said I'm not sure that in the end we simply should not disable them.
cc @lrineau

maxGimeno commented:
there is also an error on windows:
C:\CGAL\test\CGAL-4.14-Ic-61\cmake\platforms\x64_Cygwin-Windows10_MSVC2013-Release-64bits\test\Polyhedron_Demo\Plugins\Surface_mesh_approximation\surface_mesh_approximation_plugin_autogen\mocs_compilation.cpp : fatal error C1083: Cannot open compiler generated file: 'CMakeFiles\surface_mesh_approximation_plugin.dir\surface_mesh_approximation_plugin_autogen\mocs_compilation.cpp.obj': No such file or directory

maxGimeno commented:
The object file directory    C:/CGAL/test/CGAL-4.14-Ic-61/cmake/platforms/x64_Cygwin-Windows10_MSVC2013-Release-64bits/test/Polyhedron_Demo/Plugins/Surface_mesh_approximation/CMakeFiles/surface_mesh_approximation_plugin.dir/  has 195 characters. 

The name of the plugin and/or its directory is too long for Windows and it might be the cause of the errors in MSVC.

maxGimeno commented:
Also, from the compilation of the doc with Doxygen master :
/home/cgal-testsuite/cgal_doc_build/CGAL-4.14-Ic-61/include/CGAL/Variational_shape_approximation.h:976: warning: Member anchor_points(boost::param_not_found) const (function) of class CGAL::Variational_shape_approximation is not documented.
/home/cgal-testsuite/cgal_doc_build/CGAL-4.14-Ic-61/include/CGAL/Variational_shape_approximation.h:1001: warning: Member indexed_triangles(boost::param_not_found) const (function) of class CGAL::Variational_shape_approximation is not documented.

aboudev commented:
The removed named parameter `min_error_drop` of function `approximate_triangle_mesh` is used for `VERBOSE` parameter report and forwarded to the `initialize_seeds` as part of the named parameter `np`.

sloriot commented:
> The removed named parameter `min_error_drop` of function `approximate_triangle_mesh` is used for `VERBOSE` parameter report and forwarded to the `initialize_seeds` as part of the named parameter `np`.

Oh my bad! did you restore its documentation?

aboudev commented:
I have restored it in 7f7360aa44802b7ee77eaa414020d31d5c1c1352.

maxGimeno commented:
Successfully Tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-63.shtml

lrineau commented:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-63.shtml has only 23 columns for now!

lrineau commented:
Well, 26, but important platforms results are still missing.

maxGimeno commented:
Well, only remaining problems were found on MSVC 2013, and it is not there anymore, so I figured it would be enough.

lrineau commented:
The last commits of the pull-requests were a day ago. So any test that is invalid.

Plus: you said the remaining issues were with MSVC 2013... but 2017 had the same issue. And that might hide another error, or warnings.

maxGimeno commented:
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-67.shtml

MaelRL commented:
Small reminder that it would be nice if someone wrote a couple of lines (+main page image) about it on the website once it's integrated.

sloriot commented:
@lrineau I've pushed the update of the change log. I don't think there is anything now blocking the merge.

sloriot commented:
Successfully tested in CGAL-4.14-Ic-87. I'm in favor of merging this PR. @lrineau ?

maxGimeno commented:
Ther eare conflicts with master now.

lrineau commented:
I have fixed the conflict. And re-added "Under testing".

maxGimeno commented:
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-101.shtml

maxGimeno commented:
Conflicts again. 
They are trivial to fix, I fixed them every night and appart from that, 
this PR was uccessfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-101.shtml
